### PR TITLE
Improve email domain discovery page to incorporate new configuration for enabling email domain discovery for self-registration

### PR DESCRIPTION
**Pull request by Yasasr1**
### Purpose
Following improvements will be made to the existing email domain discovery page 
- Generalize the page name to "Organization Discovery"
- Move down the assign email domain button from top right corner to separate email domain discovery related configs from email domain mappings for organizations
- Improve the info message at the top to mention the flows that restricts email domains when email domain discovery is enabled
- Add a checkbox to incorporate the new config: email domain based organization discovery for self-registration

Screenshot 2024-11-05 at 10 50 23 AM

### Related issue
- https://github.com/wso2/product-is/issues/21610

### CODE REVIEW

File: features/admin.organization-discovery.v1/components/discoverable-organizations-list-layout.tsx [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+export interface DiscoverableOrganizationsListLayoutPropsInterface extends IdentifiableComponentInterface {
+    discoverableOrganizations: OrganizationListWithDiscoveryInterface
+    listItemLimit: number;
+    isDiscoverableOrganizationsFetchRequestLoading: boolean;
+    featureConfig: FeatureConfigInterface;
+    searchQuery: string;
+    onListItemLimitChange: (limit: number) => void;
+    onListOffsetChange: (offset: number) => void;
+    onSearchQueryChange: (query: string) => void;
+
```
Comment (pavinduLakshan) [inline features/admin.organization-discovery.v1/components/discoverable-organizations-list-layout.tsx:54]:
```suggestion
```

File: features/admin.organization-discovery.v1/api/use-get-organization-discovery-config.ts [hunk 2]
```diff
@@ -36,7 +36,8 @@ import {
  * @returns SWR response object containing the data, error, isValidating, mutate.
  */
 const useGetOrganizationDiscoveryConfig = <
-    Data = OrganizationDiscoveryConfigInterface & { isOrganizationDiscoveryEnabled: boolean },
+    Data = OrganizationDiscoveryConfigInterface & { isOrganizationDiscoveryEnabled: boolean,
```
Comment (Yasasr1) [inline features/admin.organization-discovery.v1/api/use-get-organization-discovery-config.ts:39]: Add these to the OrganizationDiscoveryConfigInterface

File: features/admin.organization-discovery.v1/api/use-get-organization-discovery-config.ts [hunk 3]
```diff
@@ -53,13 +54,18 @@ const useGetOrganizationDiscoveryConfig = <
     });
 
     let isOrganizationDiscoveryEnabled: boolean = false;
+    let isEmailDomainBasedSelfRegistrationEnabled: boolean = false;
 
     if ((data as OrganizationDiscoveryConfigInterface)?.properties) {
         (data as OrganizationDiscoveryConfigInterface).properties?.forEach(
             (property: OrganizationDiscoveryConfigPropertyInterface) => {
-                if (property.key === "emailDomain.enable") {
+                if (property.key === "emailDomain.enable" && property.value === "true") {
```
Comment (Yasasr1) [inline features/admin.organization-discovery.v1/api/use-get-organization-discovery-config.ts:62]: Add as constants

File: features/admin.organization-discovery.v1/components/discoverable-organizations-list-layout.scss [hunk 4]
```diff
@@ -0,0 +1,15 @@
+.discoverable-organizations-list-layout {
```
Comment (Yasasr1) [inline features/admin.organization-discovery.v1/components/discoverable-organizations-list-layout.scss:1]: Add license header

File: features/admin.organization-discovery.v1/components/discoverable-organizations-list-layout.tsx (hunk 1 above)
Comment (Yasasr1) [inline features/admin.organization-discovery.v1/components/discoverable-organizations-list-layout.tsx:54]: remove line

File: features/admin.organization-discovery.v1/components/discoverable-organizations-list-layout.tsx [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+            />
+        </ListLayout>
+    );
+};
+
+/**
+ * Default props for the component.
+ */
+DiscoverableOrganizationsListLayout.defaultProps = {
+    "data-componentid": "discoverable-organizations-list-layout"
```
Comment (Yasasr1) [inline features/admin.organization-discovery.v1/components/discoverable-organizations-list-layout.tsx:253]: refactor

File: features/admin.organization-discovery.v1/pages/organization-discovery-domains-page.tsx [hunk 6]
```diff
@@ -97,27 +79,13 @@ const OrganizationDiscoveryDomainsPage: FunctionComponent<OrganizationDiscoveryD
...
-        () =>
-            !hasRequiredScopes(
-                featureConfig?.organizationDiscovery,
-                featureConfig?.organizationDiscovery?.scopes?.update,
-                allowedScopes
-            ),
-        [ featureConfig, allowedScopes ]
-    );
+    const isEmailDomainDiscoveryForSelfRegFeatureEnabled: boolean = !featureConfig?.organizationDiscovery?.
+        disabledFeatures?.includes("organizationDiscovery.emailDomainDiscoveryForSelfReg");
```
Comment (Yasasr1) [inline features/admin.organization-discovery.v1/pages/organization-discovery-domains-page.tsx:83]: use FEATURE_DICTIONARY constant

File: features/admin.organization-discovery.v1/pages/organization-discovery-domains-page.tsx [hunk 7]
```diff
@@ -97,27 +79,13 @@ const OrganizationDiscoveryDomainsPage: FunctionComponent<OrganizationDiscoveryD
...
-    const isReadOnly: boolean = useMemo(
-        () =>
-            !hasRequiredScopes(
-                featureConfig?.organizationDiscovery,
-                featureConfig?.organizationDiscovery?.scopes?.update,
-                allowedScopes
-            ),
-        [ featureConfig, allowedScopes ]
-    );
+    const isEmailDomainDiscoveryForSelfRegFeatureEnabled: boolean = !featureConfig?.organizationDiscovery?.
```
Comment (pavinduLakshan) [inline features/admin.organization-discovery.v1/pages/organization-discovery-domains-page.tsx:82]: use isFeatureEnabled from core module

File: features/admin.organization-discovery.v1/pages/organization-discovery-domains-page.tsx [hunk 8]
```diff
@@ -140,7 +107,118 @@ const OrganizationDiscoveryDomainsPage: FunctionComponent<OrganizationDiscoveryD
...
 
-    const { isOrganizationDiscoveryEnabled } = organizationDiscoveryConfig;
+    const { isOrganizationDiscoveryEnabled, isEmailDomainBasedSelfRegistrationEnabled } = organizationDiscoveryConfig;
+    const [ isSelfRegEnabled, setIsSelfRegEnabled ] = useState<boolean>(false);
+
+    /**
+     * Fetch the self-registration configuration.
+     */
+    useEffect(() => {
+        getConnectorDetails(ServerConfigurationsConstants.USER_ONBOARDING_CONNECTOR_ID,
```
Comment (Yasasr1) [inline features/admin.organization-discovery.v1/pages/organization-discovery-domains-page.tsx:117]: use a swr hook to fetch the properties

File: features/admin.organization-discovery.v1/pages/organization-discovery-domains-page.tsx [hunk 9]
```diff
@@ -140,7 +107,118 @@ const OrganizationDiscoveryDomainsPage: FunctionComponent<OrganizationDiscoveryD
...
+    }, []);
+
+    const handleEmailDomainBasedSelfRegistration = (value: boolean): void => {
+        const updateData: OrganizationDiscoveryConfigInterface = {
+            properties: []
+        };
+
+        updateData.properties.push({
+            key: "emailDomain.enable",
+            value: isOrganizationDiscoveryEnabled ? "true" : "false"
```
Comment (Yasasr1) [inline features/admin.organization-discovery.v1/pages/organization-discovery-domains-page.tsx:168]: convert to string. check other places

**Review by pavinduLakshan — APPROVED** [review]

File: modules/i18n/src/translations/en-US/portals/organization-discovery.ts [hunk 10]
```diff
@@ -174,5 +195,10 @@ export const organizationDiscovery: organizationDiscoveryNS = {
             title: "Assign Email Domain"
         }
     },
-    title: "Email Domain Discovery"
+    selfRegistration: {
+        label: "Email domain discovery for self-registration",
+        labelHint: "when enabled, users will be self-registered in child organizations via shared applications",
```
Comment (Yasasr1) [inline modules/i18n/src/translations/en-US/portals/organization-discovery.ts:200]: w should be capital

File: features/admin.server-configurations.v1/api/governance-connectors.ts [hunk 11]
```diff
@@ -108,6 +108,42 @@ export const useGovernanceConnectors = <
...
 };
 
+/**
+ * Get details of a governance connector.
+ *
+ * @param categoryId - ID of the connector category
+ * @param connectorId - ID of the connector
+ * @returns the governance connector.
+ */
+export const useGovernanceConnectorDetails = <
```
Comment (pavinduLakshan) [inline features/admin.server-configurations.v1/api/governance-connectors.ts:118]: We follow the naming convention `useGet + resource name` for SWR hooks. Shall we follow the same here?

```suggestion
export const useGetGovernanceConnectorById = <
```

File: .changeset/shiny-bulldogs-bathe.md [hunk 12]
```diff
@@ -0,0 +1,6 @@
+---
+"@wso2is/admin.organization-discovery.v1": patch
+"@wso2is/i18n": patch
+---
```
Comment (pavinduLakshan) [inline .changeset/shiny-bulldogs-bathe.md:4]: Let's add a separate changeset with `minor` update for server-configurations package too.

File: modules/i18n/src/translations/en-US/portals/organization-discovery.ts [hunk 13]
```diff
@@ -118,6 +120,16 @@ export const organizationDiscovery: organizationDiscoveryNS = {
                 message: "Validating unsuccessful"
             }
         },
+        disableEmailDomainBasedSelfRegistration: {
+            error: {
+                description: "An error occurred while disabling email domain discovery for self-registration.",
+                message: "Disabling unsuccessful"
```
Comment (pavinduLakshan) [inline modules/i18n/src/translations/en-US/portals/organization-discovery.ts:126]:
```suggestion
                message: "Disabling failed."
```

**Review by pavinduLakshan — DISMISSED** [review]

**Pull request merged**